File: Projects/SwiftPerfLab/Core/Generator/MapTestDataGenerator.swift
import Foundation
import CoreLocation

/// Ensures Swift 6 concurrency compliance.
/// CLLocationCoordinate2D is a thread-safe C-struct, safely bypassing strict concurrency checks.
extension CLLocationCoordinate2D: @retroactive @unchecked Sendable {}

/// A thread-safe namespace for generating spatial datasets.
///
/// Architected as a stateless enum instead of an actor or class to avoid 
/// asynchronous context-switching overhead, achieving maximum CPU throughput.
enum MapTestDataGenerator {
    
    /// Generates a spatially distributed array of map annotations using a sine wave distribution. 
    /// This provides a baseline for testing map rendering and spatial indexing performance.
    ///
    /// - Parameter count: The total number of map points to generate. Defaults to 20.
    /// - Returns: An array of initialized `MapPoint` entities.
    nonisolated static func generatePoints(count: Int = 20) -> [MapPoint] {
        // Base coordinates anchored to Istanbul for spatial rendering tests.
        let centerLat = 41.0082
        let centerLon = 28.9784
        let spread = 0.05

        // Generates coordinates using trigonometric distribution.
        // Using .map on a Range pre-allocates exact memory, avoiding CPU overhead 
        // from dynamic array resizing.
        return (0..<count).map { i in
            let progress = Double(i) / Double(count)
            
            // Calculates linear latitude and sine-wave longitude oscillation.
            let latOffset = (progress - 0.5) * spread * 2
            let lonOffset = sin(progress * .pi * 4) * spread

            return MapPoint(
                // Uses a deterministic integer ID to eliminate CPU and RAM overhead 
                // of UUID() generation.
                id: i,
                coordinate: CLLocationCoordinate2D(
                    latitude: centerLat + latOffset,
                    longitude: centerLon + lonOffset
                ),
                title: "Pin \(i + 1)"
            )
        }
    }
}

File: Projects/KotlinMultiPlatformPerfLab/composeApp/src/commonMain/kotlin/com/tez/perflab/generators/MapTestDataGenerator.kt
package com.tez.perflab.generators

import com.tez.perflab.models.MapPoint
import kotlin.math.PI
import kotlin.math.sin

/**
 * A stateless, thread-safe namespace for generating deterministic spatial datasets.
 * Creates standardized baselines for testing rendering performance and spatial
 * indexing across mapping engines (e.g., MapKit vs. Google Maps).
 *
 * Executes synchronously to eliminate the CPU overhead of coroutine context
 * switching, ensuring accurate performance measurements.
 */
object MapTestDataGenerator {

    /**
     * Generates a spatially distributed array of map annotations using trigonometric distribution.
     * Executes synchronously to prioritize mathematical throughput over thread dispatching latency.
     *
     * @param count The total number of map points to generate. Defaults to 20.
     * @return A pre-allocated list of initialized MapPoint entities.
     */
    fun generatePoints(count: Int = 20): List<MapPoint> {

        // Base coordinates anchored to Istanbul for localized spatial rendering tests.
        val centerLat = 41.0082
        val centerLon = 28.9784
        val spread = 0.05

        // Default zoom level to synchronize map scales across platforms.
        val defaultZoom = 14f

        // Explicitly instantiating a fixed-size List circumvents the CPU overhead
        // of dynamic array resizing, isolating the benchmark to mathematical throughput.
        return List(count) { i ->
            val progress = i.toDouble() / count.toDouble()

            // Calculates linear latitude progression and sine-wave based longitude oscillation.
            val latOffset = (progress - 0.5) * spread * 2.0
            val lonOffset = sin(progress * PI * 4.0) * spread

            MapPoint(
                // Uses integer IDs to eliminate UUID generation overhead during benchmarking.
                id = i,
                latitude = centerLat + latOffset,
                longitude = centerLon + lonOffset,
                title = "Pin ${i + 1}",
                zoomLevel = defaultZoom
            )
        }
    }
}